src/server/CollisionDetector.hpp

```diff
@@ -3,12 +3,28 @@
 
 #include <stdio.h>
 #include "Group.hpp"
+#include "../common/util.hpp"
+#include "../common/GroupShape.hpp"
 
 class CollisionDetector
 {
     public:
         CollisionDetector();
-        void detectGroupCollisions(std::vector<Group*> groups);
+        template<typename T> void detectAndHandleShapeCollisions(const std::vector<T*>& shapes);
+        void handleGroupShapeCollision(GroupShape* group_a, GroupShape* group_b);
+        void detectAndHandleGroupShapeCollisoin(std::vector<GroupShape*> group_shapes);
 };
 
+// TODO: Move this into a tpp file?
+template<class Group> void CollisionDetector::detectAndHandleShapeCollisions(const std::vector<Group*>& shapes) {
```

Comment:
this line is confusing af

Reply:
the casting feels unnecessary. seems like renderable state like color should be wrapped around the physics object. like I think the client can have a RenderableGroup with a reference to a PhysicsGroup, the collision detector only care about PhysicsGroups, so we only ever pass it those. no type casting or weird templates